packages/core/src/extension/extension-service.ts

```diff
@@ -0,0 +1,58 @@
+import { noop } from 'lodash';
+
+import { IframeEventListener, IframeEventPoster } from '../common/iframe';
+
+import {
+    ExtensionListenEventMap,
+    ExtensionListenEventType,
+    ExtensionPostEvent,
+    ExtensionPostEventType,
+} from './extension-client';
+
+export default class ExtensionService {
+    private _extensionId?: string;
+
+    constructor(
+        private _eventListener: IframeEventListener<ExtensionListenEventMap>,
+        private _eventPoster: IframeEventPoster<ExtensionPostEvent>,
+    ) {
+        this._eventPoster.setTarget(window.parent);
+    }
+
+    initialize(extensionId: string) {
+        if (!extensionId) {
+            throw new Error('Extension Id not found.');
+        }
+
+        this._extensionId = extensionId;
+
+        this._eventListener.listen();
+    }
+
+    post(event: ExtensionPostEvent): void {
+        if (!this._extensionId) {
+            return;
+        }
+
+        if (!Object.values(ExtensionPostEventType).includes(event.type)) {
+            throw new Error(`${event.type} is not supported.`);
+        }
+
+        const payload = {
+            ...event.payload,
+            extensionId: this._extensionId,
+        };
+
+        this._eventPoster.post({ ...event, payload });
+    }
+
+    addListener(eventType: ExtensionListenEventType, callback: () => void = noop): () => void {
+        if (!Object.values(ExtensionListenEventType).includes(eventType)) {
+            return noop;
```

Comment:
Should it raise an exception instead of failing silently?

Reply:
This is on the client side so I am not sure if we throw an exception? Probably makes sense to throw an exception to improve the developer experience. Will change it accordingly :D